compiler/rustc_lexer/src/unescape.rs

```diff
@@ -90,6 +93,104 @@ where
         Mode::RawStr | Mode::RawByteStr => {
             unescape_raw_str_or_raw_byte_str(src, mode == Mode::RawByteStr, callback)
         }
+        Mode::CStr | Mode::RawCStr => unreachable!(),
+    }
+}
+
+pub enum CStrUnit {
+    Byte(u8),
+    Char(char),
+}
+
+pub fn unescape_c_string<F>(src: &str, mode: Mode, callback: &mut F)
```

Comment:
Sorry, I don't understand what you are saying.
Both byte and C strings support non-UTF8 so (Rust) `char`s are out of the question.
I'm concerned about the difference between byte strings and C strings, both produce arbitrary non-UTF `[u8]` and any differences between them should eventually be eliminated (that's the point of https://github.com/rust-lang/rfcs/pull/3349 from what I remember).

Reply:
you are true. but currently, both byte strings and normal strings emit `char`s in their implementation. Byte strings just use the codepoints to represent the byte values, but that would need to be changed to an enum if we were to implement that rfc.

Reply:
My understanding is that most of this complication comes from the fact that the C-str RFC explicit states that it supports both `\u` and `\x` escapes in `c""` literals. Is that correct?

Reply:
@compiler-errors  Not necessarily about the `\u` escape, but more about the `\x` escape which has a different meaning in byte strings and characters. nnethercote [comment](https://github.com/rust-lang/rfcs/pull/3349#issuecomment-1334667652) at the RFC mentioned above suggested that a table should make this clearer:

  | Example | # sets* | Characters | Escapes
-- | -- | -- | -- | --
Character | 'H' | 0 | All Unicode | Quote & ASCII & Unicode
String | "hello" | 0 | All Unicode | Quote & ASCII & Unicode
Raw string | r#"hello"# | <256 | All Unicode | N/A
Byte | b'H' | 0 | All ASCII | Quote & Byte
Byte string | b"hello" | 0 | All ASCII | Quote & Byte
Raw byte string | br#"hello"# | <256 | All ASCII | N/A
**C string** | c"hello" | 0 | All bytes except NUL | **Quote & Byte & Unicode**

Note that since normal strings accept unicode, we can emit `char`s that correspond to the actual characters. But for byte strings this is different. Byte strings allow bytes that are not encoded as UTF-8. (e.g. `\xFF` allowed in byte strings but not in normal strings) How do we unescape them currently? We currently emit the **codepoint** (e.g. `\xFF` -> `ÿ` U+00FF) for byte strings and then interpret the values later on.

That means that `ÿ` character emitted by a normal string means "ÿ", with codepoint U+00FF, encoded in UTF-8 as `0xC3 0xBF`. But this emitted for a byte string would mean the byte `0xFF` only. C strings are explicitly allowed to have both, therefore it is necessary to use an enum to convey either the character encoded as UTF-8 or the byte value.

Reply:
@fee1-dead In the entry in the "Characters"  column, in the "C string" row, do you really mean "all bytes exept NUL"?.  IIRC Rust files are required to be valid UTF-8, and RFC 3348 has changed nothing about that. At least I found nothing in the RFC's text indicating that. The goal was more about the escapes column: the encoded *result* can be a non-valid unicode string, but the literal itself still has to be valid UTF-8. Otherwise this would mean that programs processing rust source code cannot assume UTF-8 validity of the source code any more. In other words, any program that uses Rust's `String` type to represent a slice of Rust code (including Rust's proc macro infrastructure!) would fail for specific snippets containing c strings that have invalid UTF-8.

I think that entry should rather read "All Unicode" or "All UTF-8".

Reply:
@est31: corrected, thanks.